# A similar fix will be merged to a future version of mbedTLS, please track
# progress via issue https://github.com/ARMmbed/mbedtls/issues/3141.

diff --git a/library/ssl_tls.c b/library/ssl_tls.c
index 9bb9dc23cde94a23bb71a2d421410a7cea039816..6e7926a99220e79935396cbc2b7d3f22720bf906 100644
--- a/library/ssl_tls.c
+++ b/library/ssl_tls.c
@@ -4250,12 +4250,14 @@ int mbedtls_ssl_context_save( mbedtls_ssl_context *ssl,
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "There is pending outgoing data" ) );
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
     }
+#if 0
     /* Protocol must be DTLS, not TLS */
     if( ssl->conf->transport != MBEDTLS_SSL_TRANSPORT_DATAGRAM )
     {
         MBEDTLS_SSL_DEBUG_MSG( 1, ( "Only DTLS is supported" ) );
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
     }
+#endif
     /* Version must be 1.2 */
     if( ssl->tls_version != MBEDTLS_SSL_VERSION_TLS1_2 )
     {
@@ -4401,6 +4403,16 @@ int mbedtls_ssl_context_save( mbedtls_ssl_context *ssl,
     }
 #endif /* MBEDTLS_SSL_ALPN */
 
+    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_STREAM )
+    {
+        used += 8;
+        if( used <= buf_len )
+        {
+            memcpy( p, ssl->in_ctr, 8 );
+            p += 8;
+        }
+    }
+
     /*
      * Done
      */
@@ -4411,7 +4423,19 @@ int mbedtls_ssl_context_save( mbedtls_ssl_context *ssl,
 
     MBEDTLS_SSL_DEBUG_BUF( 4, "saved context", buf, used );
 
+#if 0
+    /* At the moment of fork (when we call this function to serialize the TLS
+     * context and send to the child), we don't know (1) whether the child
+     * process will use it at all -- maybe both TLS endpoints will be used by
+     * the parent process, and (2) which TLS endpoint will be closed and which
+     * endpoint will be used. Thus, we must not reset the session since it may
+     * be continued to be used.
+     * Currently we are relying on the application to be "sane" and not use
+     * the same endpoint in two different processes. */
     return( mbedtls_ssl_session_reset_int( ssl, 0 ) );
+#else
+    return( 0 );
+#endif
 }
 
 /*
@@ -4449,7 +4473,10 @@ static int ssl_context_load( mbedtls_ssl_context *ssl,
      * We can't check that the config matches the initial one, but we can at
      * least check it matches the requirements for serializing.
      */
+#if 0
     if( ssl->conf->transport != MBEDTLS_SSL_TRANSPORT_DATAGRAM ||
+#else
+    if(
         ssl->conf->max_tls_version < MBEDTLS_SSL_VERSION_TLS1_2 ||
         ssl->conf->min_tls_version > MBEDTLS_SSL_VERSION_TLS1_2 ||
 #if defined(MBEDTLS_SSL_RENEGOTIATION)
@@ -4459,6 +4486,7 @@ static int ssl_context_load( mbedtls_ssl_context *ssl,
     {
         return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
     }
+#endif
 
     MBEDTLS_SSL_DEBUG_BUF( 4, "context to load", buf, len );
 
@@ -4670,6 +4698,15 @@ static int ssl_context_load( mbedtls_ssl_context *ssl,
     ssl->in_epoch = 1;
 #endif
 
+    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_STREAM )
+    {
+        if( (size_t)( end - p ) < 8 )
+            return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+        memcpy( ssl->in_ctr, p, 8 );
+        p += 8;
+    }
+
     /* mbedtls_ssl_reset() leaves the handshake sub-structure allocated,
      * which we don't want - otherwise we'd end up freeing the wrong transform
      * by calling mbedtls_ssl_handshake_wrapup_free_hs_transform()
diff --git a/library/Makefile b/library/Makefile
index dd16d061588321952a74944a9a0e17427cd7544e..680b67b52cfe4acf686e77eb83ddfc58bcfac05e 100644
--- a/library/Makefile
+++ b/library/Makefile
@@ -51,6 +51,8 @@ SOEXT_TLS?=so.19
 SOEXT_X509?=so.4
 SOEXT_CRYPTO?=so.13
 
+SUFFIX ?=
+
 # Set AR_DASH= (empty string) to use an ar implementation that does not accept
 # the - prefix for command line options (e.g. llvm-ar)
 AR_DASH ?= -
@@ -199,10 +201,11 @@ else
 all: shared static
 endif
 
-static: libmbedcrypto.a libmbedx509.a libmbedtls.a
+static: libmbedcrypto$(SUFFIX).a libmbedx509$(SUFFIX).a libmbedtls$(SUFFIX).a
 	cd ../tests && echo "This is a seedfile that contains 64 bytes (65 on Windows)......" > seedfile
 
-shared: libmbedcrypto.$(DLEXT) libmbedx509.$(DLEXT) libmbedtls.$(DLEXT)
+shared: libmbedcrypto$(SUFFIX).$(DLEXT) libmbedx509$(SUFFIX).$(DLEXT) libmbedtls$(SUFFIX).$(DLEXT) \
+		libmbedcrypto$(SUFFIX).$(SOEXT_CRYPTO) libmbedx509$(SUFFIX).$(SOEXT_X509) libmbedtls$(SUFFIX).$(SOEXT_TLS)
 
 # Windows builds under Mingw can fail if make tries to create archives in the same
 # directory at the same time - see https://bugs.launchpad.net/gcc-arm-embedded/+bug/1848002.
@@ -213,7 +216,7 @@ libmbedx509.a: | libmbedcrypto.a
 endif
 
 # tls
-libmbedtls.a: $(OBJS_TLS)
+libmbedtls%.a: $(OBJS_TLS)
 	echo "  AR    $@"
 	$(AR) $(ARFLAGS) $@ $(OBJS_TLS)
 ifdef APPLE_BUILD
@@ -223,12 +226,12 @@ ifneq ($(APPLE_BUILD),0)
 endif
 endif
 
-libmbedtls.$(SOEXT_TLS): $(OBJS_TLS) libmbedx509.so
+libmbedtls%.$(SOEXT_TLS): $(OBJS_TLS) libmbedx509%.so
 	echo "  LD    $@"
-	$(CC) -shared -Wl,-soname,$@ -o $@ $(OBJS_TLS) -L. -lmbedx509 -lmbedcrypto $(LOCAL_LDFLAGS) $(LDFLAGS)
+	$(CC) -shared -Wl,-soname,$@ -o $@ $(OBJS_TLS) -L. -lmbedx509$(SUFFIX) -lmbedcrypto$(SUFFIX) $(LOCAL_LDFLAGS) $(LDFLAGS)
 
 ifneq ($(SOEXT_TLS),so)
-libmbedtls.so: libmbedtls.$(SOEXT_TLS)
+libmbedtls%.so: libmbedtls%.$(SOEXT_TLS)
 	echo "  LN    $@ -> $<"
 	ln -sf $< $@
 endif
@@ -242,7 +245,7 @@ libmbedtls.dll: $(OBJS_TLS) libmbedx509.dll
 	$(CC) -shared -Wl,-soname,$@ -Wl,--out-implib,$@.a -o $@ $(OBJS_TLS) -lws2_32 -lwinmm -lgdi32 -L. -lmbedx509 -lmbedcrypto -static-libgcc $(LOCAL_LDFLAGS) $(LDFLAGS)
 
 # x509
-libmbedx509.a: $(OBJS_X509)
+libmbedx509%.a: $(OBJS_X509)
 	echo "  AR    $@"
 	$(AR) $(ARFLAGS) $@ $(OBJS_X509)
 ifdef APPLE_BUILD
@@ -252,12 +255,12 @@ ifneq ($(APPLE_BUILD),0)
 endif
 endif
 
-libmbedx509.$(SOEXT_X509): $(OBJS_X509) libmbedcrypto.so
+libmbedx509%.$(SOEXT_X509): $(OBJS_X509) libmbedcrypto%.so
 	echo "  LD    $@"
-	$(CC) -shared -Wl,-soname,$@ -o $@ $(OBJS_X509) -L. -lmbedcrypto $(LOCAL_LDFLAGS) $(LDFLAGS)
+	$(CC) -shared -Wl,-soname,$@ -o $@ $(OBJS_X509) -L. -lmbedcrypto$(SUFFIX) $(LOCAL_LDFLAGS) $(LDFLAGS)
 
 ifneq ($(SOEXT_X509),so)
-libmbedx509.so: libmbedx509.$(SOEXT_X509)
+libmbedx509%.so: libmbedx509%.$(SOEXT_X509)
 	echo "  LN    $@ -> $<"
 	ln -sf $< $@
 endif
@@ -271,7 +274,7 @@ libmbedx509.dll: $(OBJS_X509) libmbedcrypto.dll
 	$(CC) -shared -Wl,-soname,$@ -Wl,--out-implib,$@.a -o $@ $(OBJS_X509) -lws2_32 -lwinmm -lgdi32 -L. -lmbedcrypto -static-libgcc $(LOCAL_LDFLAGS) $(LDFLAGS)
 
 # crypto
-libmbedcrypto.a: $(OBJS_CRYPTO)
+libmbedcrypto%.a: $(OBJS_CRYPTO)
 	echo "  AR    $@"
 	$(AR) $(ARFLAGS) $@ $(OBJS_CRYPTO)
 ifdef APPLE_BUILD
@@ -281,12 +284,12 @@ ifneq ($(APPLE_BUILD),0)
 endif
 endif
 
-libmbedcrypto.$(SOEXT_CRYPTO): $(OBJS_CRYPTO)
+libmbedcrypto%.$(SOEXT_CRYPTO): $(OBJS_CRYPTO)
 	echo "  LD    $@"
 	$(CC) -shared -Wl,-soname,$@ -o $@ $(OBJS_CRYPTO) $(LOCAL_LDFLAGS) $(LDFLAGS)
 
 ifneq ($(SOEXT_CRYPTO),so)
-libmbedcrypto.so: libmbedcrypto.$(SOEXT_CRYPTO)
+libmbedcrypto%.so: libmbedcrypto%.$(SOEXT_CRYPTO)
 	echo "  LN    $@ -> $<"
 	ln -sf $< $@
 endif
 
